src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -339,12 +340,301 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+    // The second perturbation vector still can't comprise of more than one
+    // change in the bin codes. This is because of the way perturbation vectors
+    // are generated: First we create the one with the smallest score (Ao) and
+    // then we either add 1 extra dimension to it (Ae) or shift it by one (As).
+    // Since As contains the second smallest score, and Ae contains both the
+    // smallest and the second smallest, it's obvious that score(Ae) >
+    // score(As). Therefore the second perturbation vector is ALWAYS the vector
+    // containing only the second-lowest scoring perturbation.
+    
+    
+    double minscore2 = scores[0];
+    size_t minloc2 = 0;
+    for (size_t s = 0; s < 2 * numProj; ++s) // here we can't start from 1
+    {
+      if ( minscore2 > scores[s] && s != minloc) //second smallest
+      {
+        minscore2 = scores[s];
+        minloc2 = s;
+      }
+    }
+
+    // add or subtract 1 to create second-lowest scoring vector
+    additionalProbingBins(positions[minloc2], 1) += actions[minloc2];
+    return;
+    
+  }
+
+  // sort everything in increasing order
+  arma::Col<long long unsigned int> sortidx = arma::sort_index(scores);
+  scores = scores(sortidx);
+  actions = actions(sortidx);
+  positions = positions(sortidx);
+
+
+  // Theory:
+  // From the paper: This is the part that creates the probing sequence.
+  // A probing sequence is a sequence of T probing bins where a query's
+  // neighbors are most likely to be. Likelihood is dependent only on a bin's
+  // score, which is the sum of scores of all dimension-action pairs, so we
+  // need to calculate the T smallest sums of scores that are not conflicting.
+  //
+  // Method:
+  // Store each perturbation set (pair of (dimension, action)) in a
+  // std::vector. Create a minheap of scores, with each node pointing to its
+  // relevant perturbation set. Each perturbation set popped from the minheap
+  // is the next most likely perturbation set.
+  // Transform perturbation set to perturbation vector by setting the
+  // dimensions specified by the set to queryCode+action (action is {-1, 1}).
+
+  std::vector<size_t> Ao;
+  Ao.push_back(0); // initial perturbation holds smallest score (0 if sorted)
+
+  std::vector< std::vector<size_t> > perturbationSets;
+  perturbationSets.push_back(Ao); // storage of perturbation sets
+
+
+  // define a priority queue with CompareGreater as a minheap
+  std::priority_queue<
+    std::pair<double, size_t>,        // contents: pairs of (score, index)
+    std::vector<                      // container: vector of pairs
+      std::pair<double, size_t>
+      >,
+    mlpack::neighbor::CompareGreater // comparator of pairs(compare scores)
+  > minHeap; // our minheap
+
+  // Start by adding the lowest scoring set to the minheap
+  std::pair<double, size_t> pair0( perturbationScore(Ao, scores), 0 );
+  minHeap.push(pair0);
+
+  double prevScore = 0; // store score of smallest inserted vector (for assert)
```

Comment:
If this variable is here just for an assert, then I think once the tests show the algorithm is working then we can remove this.